ICE: invalid lit/ty combination in `lit_to_mir_constant`

<!--
[31mICE[0m: Rustc ./3AA84AE74FF1EE67A8A4F3F02E9F51C9BA41787B2D9189CD8456EB29B8F12B65.rs '' 'error: internal compiler error: compiler/rustc_mir_build/src/builder/expr/as_constant.rs:162:14: invalid lit/ty combination in `lit_to_mir_constant`: Int(Pu128(3), Unsuffixed): u8', 'error: internal compiler error: compiler/rustc_mir_build/src/builder/expr/as_constant.rs:162:14: invalid lit/ty combination in `lit_to_mir_constant`: Int(Pu128(3), Unsuffixed): u8'
File: /tmp/im/3AA84AE74FF1EE67A8A4F3F02E9F51C9BA41787B2D9189CD8456EB29B8F12B65.rs
-->

auto-reduced (treereduce-rust):
````rust




enum test { thing = -5 >> 1_usize }
pub fn main(x: impl Iterator<Item = Foo>) {
    assert_eq!(test::thing as u8, -3);
}

````

original:
````rust
//@ run-pass
#![feature(inherent_associated_types, auto_traits, negative_impls)]


enum test { thing = -5 >> 1_usize }
pub fn main(x: impl Iterator<Item = Foo>) {
    assert_eq!(test::thing as u8, -3);
}

````

Version information
````
rustc 1.93.0-nightly (f15a7f385 2025-11-04)
binary: rustc
commit-hash: f15a7f38580ddbdc1a23909dd05cf6cc6d9f3919
commit-date: 2025-11-04
host: x86_64-unknown-linux-gnu
release: 1.93.0-nightly
LLVM version: 21.1.3

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/f15a7f38580ddbdc1a23909dd05cf6cc6d9f3919/compiler/rustc_mir_build/src/builder/expr/as_constant.rs#L156-L168

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0412]: cannot find type `Foo` in this scope
 --> /tmp/icemaker_global_tempdir.3332Twemxj1F/rustc_testrunner_tmpdir_reporting.qy88N4M4NQfD/mvce.rs:6:37
  |
6 | pub fn main(x: impl Iterator<Item = Foo>) {
  |                                     ^^^ not found in this scope

warning: type `test` should have an upper camel case name
 --> /tmp/icemaker_global_tempdir.3332Twemxj1F/rustc_testrunner_tmpdir_reporting.qy88N4M4NQfD/mvce.rs:5:6
  |
5 | enum test { thing = -5 >> 1_usize }
  |      ^^^^ help: convert the identifier to upper camel case: `Test`
  |
  = note: `#[warn(non_camel_case_types)]` (part of `#[warn(nonstandard_style)]`) on by default

warning: variant `thing` should have an upper camel case name
 --> /tmp/icemaker_global_tempdir.3332Twemxj1F/rustc_testrunner_tmpdir_reporting.qy88N4M4NQfD/mvce.rs:5:13
  |
5 | enum test { thing = -5 >> 1_usize }
  |             ^^^^^ help: convert the identifier to upper camel case: `Thing`

error[E0580]: `main` function has wrong type
 --> /tmp/icemaker_global_tempdir.3332Twemxj1F/rustc_testrunner_tmpdir_reporting.qy88N4M4NQfD/mvce.rs:6:1
  |
6 | pub fn main(x: impl Iterator<Item = Foo>) {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ incorrect number of function parameters
  |
  = note: expected signature `fn()`
             found signature `fn(impl Iterator<Item = Foo>)`

error: internal compiler error: compiler/rustc_mir_build/src/builder/expr/as_constant.rs:162:14: invalid lit/ty combination in `lit_to_mir_constant`: Int(Pu128(3), Unsuffixed): u8


thread 'rustc' (120936) panicked at compiler/rustc_mir_build/src/builder/expr/as_constant.rs:162:14:
Box<dyn Any>
stack backtrace:
   0:     0x7f1682e4e723 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hba09dcb7d3e45213
   1:     0x7f1683601afc - core::fmt::write::h2fd1bfda7b24e739
   2:     0x7f1682e03373 - std::io::Write::write_fmt::hafd64992b56d9630
   3:     0x7f1682e148a2 - std::sys::backtrace::BacktraceLock::print::h2cca904276f284ee
   4:     0x7f1682e1a8f9 - std::panicking::default_hook::{{closure}}::h0fd41c13a872576f
   5:     0x7f1682e1a423 - std::panicking::default_hook::hf991b196db450887
   6:     0x7f1681e22301 - std[208f11f5ffaa1bb5]::panicking::update_hook::<alloc[16665b45e0d8d179]::boxed::Box<rustc_driver_impl[b0ac0cf66d32bd3c]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x7f1682e1ad1f - std::panicking::panic_with_hook::h6dbfe896f7b6fb93
   8:     0x7f1681e5ed61 - std[208f11f5ffaa1bb5]::panicking::begin_panic::<rustc_errors[2f7fc51279732603]::ExplicitBug>::{closure#0}
   9:     0x7f1681e50e36 - std[208f11f5ffaa1bb5]::sys::backtrace::__rust_end_short_backtrace::<std[208f11f5ffaa1bb5]::panicking::begin_panic<rustc_errors[2f7fc51279732603]::ExplicitBug>::{closure#0}, !>
  10:     0x7f1681e4e3b4 - std[208f11f5ffaa1bb5]::panicking::begin_panic::<rustc_errors[2f7fc51279732603]::ExplicitBug>
  11:     0x7f1681e70191 - <rustc_errors[2f7fc51279732603]::diagnostic::BugAbort as rustc_errors[2f7fc51279732603]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  12:     0x7f16823f634a - rustc_middle[149f886374b22654]::util::bug::opt_span_bug_fmt::<rustc_span[3c70ddd04938a422]::span_encoding::Span>::{closure#0}
  13:     0x7f16823f64ca - rustc_middle[149f886374b22654]::ty::context::tls::with_opt::<rustc_middle[149f886374b22654]::util::bug::opt_span_bug_fmt<rustc_span[3c70ddd04938a422]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  14:     0x7f16823e6edb - rustc_middle[149f886374b22654]::ty::context::tls::with_context_opt::<rustc_middle[149f886374b22654]::ty::context::tls::with_opt<rustc_middle[149f886374b22654]::util::bug::opt_span_bug_fmt<rustc_span[3c70ddd04938a422]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  15:     0x7f168014e2b0 - rustc_middle[149f886374b22654]::util::bug::bug_fmt
  16:     0x7f1683a1123d - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::as_constant
  17:     0x7f1683a0d61a - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::as_rvalue
  18:     0x7f1683a22eb3 - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::expr_into_dest
  19:     0x7f1683a16687 - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::as_temp::{closure#0}
  20:     0x7f1683a1fb4e - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::expr_as_place
  21:     0x7f1683a200ba - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::expr_as_place
  22:     0x7f1683a2374c - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::expr_into_dest
  23:     0x7f1683a13349 - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::as_operand
  24:     0x7f1683a123c9 - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::as_operand
  25:     0x7f1683a0d54b - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::as_rvalue
  26:     0x7f1683a22eb3 - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::expr_into_dest
  27:     0x7f1683a16687 - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::as_temp::{closure#0}
  28:     0x7f1683a1fb4e - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::expr_as_place
  29:     0x7f1683a200ba - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::expr_as_place
  30:     0x7f1683a2561a - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::expr_into_dest
  31:     0x7f1683a16687 - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::as_temp::{closure#0}
  32:     0x7f16839fae05 - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::stmt_expr
  33:     0x7f16839facbe - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::stmt_expr
  34:     0x7f16845b9aff - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::ast_block_stmts
  35:     0x7f16845b8e6c - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::ast_block
  36:     0x7f1683a2489b - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::expr_into_dest
  37:     0x7f1683a23b44 - <rustc_mir_build[35a57232ba2ab173]::builder::Builder>::expr_into_dest
  38:     0x7f16839f00ea - rustc_mir_build[35a57232ba2ab173]::builder::build_mir
  39:     0x7f16836098e9 - rustc_mir_transform[85abc98900f95033]::mir_built
  40:     0x7f16836098a5 - rustc_query_impl[680d9d1e243271fd]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[680d9d1e243271fd]::query_impl::mir_built::dynamic_query::{closure#2}::{closure#0}, rustc_middle[149f886374b22654]::query::erase::Erased<[u8; 8usize]>>
  41:     0x7f16838e81b7 - rustc_query_system[611debe4f94c97e8]::query::plumbing::try_execute_query::<rustc_query_impl[680d9d1e243271fd]::DynamicConfig<rustc_data_structures[b0960140c6e97eca]::vec_cache::VecCache<rustc_span[3c70ddd04938a422]::def_id::LocalDefId, rustc_middle[149f886374b22654]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[611debe4f94c97e8]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[680d9d1e243271fd]::plumbing::QueryCtxt, false>
  42:     0x7f16838e7c09 - rustc_query_impl[680d9d1e243271fd]::query_impl::mir_built::get_query_non_incr::__rust_end_short_backtrace
  43:     0x7f16838e392d - rustc_mir_build[35a57232ba2ab173]::check_unsafety::check_unsafety
  44:     0x7f16838e36a3 - rustc_query_impl[680d9d1e243271fd]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[680d9d1e243271fd]::query_impl::check_unsafety::dynamic_query::{closure#2}::{closure#0}, rustc_middle[149f886374b22654]::query::erase::Erased<[u8; 0usize]>>
  45:     0x7f16838e233f - rustc_query_system[611debe4f94c97e8]::query::plumbing::try_execute_query::<rustc_query_impl[680d9d1e243271fd]::DynamicConfig<rustc_data_structures[b0960140c6e97eca]::vec_cache::VecCache<rustc_span[3c70ddd04938a422]::def_id::LocalDefId, rustc_middle[149f886374b22654]::query::erase::Erased<[u8; 0usize]>, rustc_query_system[611debe4f94c97e8]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[680d9d1e243271fd]::plumbing::QueryCtxt, false>
  46:     0x7f16838e20bd - rustc_query_impl[680d9d1e243271fd]::query_impl::check_unsafety::get_query_non_incr::__rust_end_short_backtrace
  47:     0x7f16838e11a6 - <rustc_middle[149f886374b22654]::ty::context::TyCtxt>::par_hir_body_owners::<rustc_interface[62b8c2fcefc9fa3c]::passes::run_required_analyses::{closure#1}::{closure#0}>::{closure#0}
  48:     0x7f16838deb18 - rustc_interface[62b8c2fcefc9fa3c]::passes::analysis
  49:     0x7f16838ddfb3 - rustc_query_impl[680d9d1e243271fd]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[680d9d1e243271fd]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[149f886374b22654]::query::erase::Erased<[u8; 0usize]>>
  50:     0x7f168469a94e - rustc_query_system[611debe4f94c97e8]::query::plumbing::try_execute_query::<rustc_query_impl[680d9d1e243271fd]::DynamicConfig<rustc_query_system[611debe4f94c97e8]::query::caches::SingleCache<rustc_middle[149f886374b22654]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[680d9d1e243271fd]::plumbing::QueryCtxt, false>
  51:     0x7f168469a586 - rustc_query_impl[680d9d1e243271fd]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  52:     0x7f168489eb4a - <rustc_interface[62b8c2fcefc9fa3c]::passes::create_and_enter_global_ctxt<core[61e03416e0020a9]::option::Option<rustc_interface[62b8c2fcefc9fa3c]::queries::Linker>, rustc_driver_impl[b0ac0cf66d32bd3c]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[61e03416e0020a9]::ops::function::FnOnce<(&rustc_session[6e87ae9f1f365eaa]::session::Session, rustc_middle[149f886374b22654]::ty::context::CurrentGcx, alloc[16665b45e0d8d179]::sync::Arc<rustc_data_structures[b0960140c6e97eca]::jobserver::Proxy>, &std[208f11f5ffaa1bb5]::sync::once_lock::OnceLock<rustc_middle[149f886374b22654]::ty::context::GlobalCtxt>, &rustc_data_structures[b0960140c6e97eca]::sync::worker_local::WorkerLocal<rustc_middle[149f886374b22654]::arena::Arena>, &rustc_data_structures[b0960140c6e97eca]::sync::worker_local::WorkerLocal<rustc_hir[9ca677bb78873a0a]::Arena>, rustc_driver_impl[b0ac0cf66d32bd3c]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  53:     0x7f16846fed73 - rustc_interface[62b8c2fcefc9fa3c]::interface::run_compiler::<(), rustc_driver_impl[b0ac0cf66d32bd3c]::run_compiler::{closure#0}>::{closure#1}
  54:     0x7f1684491eb8 - std[208f11f5ffaa1bb5]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[62b8c2fcefc9fa3c]::util::run_in_thread_with_globals<rustc_interface[62b8c2fcefc9fa3c]::util::run_in_thread_pool_with_globals<rustc_interface[62b8c2fcefc9fa3c]::interface::run_compiler<(), rustc_driver_impl[b0ac0cf66d32bd3c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  55:     0x7f1684491b88 - <<std[208f11f5ffaa1bb5]::thread::Builder>::spawn_unchecked_<rustc_interface[62b8c2fcefc9fa3c]::util::run_in_thread_with_globals<rustc_interface[62b8c2fcefc9fa3c]::util::run_in_thread_pool_with_globals<rustc_interface[62b8c2fcefc9fa3c]::interface::run_compiler<(), rustc_driver_impl[b0ac0cf66d32bd3c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[61e03416e0020a9]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  56:     0x7f168449302f - std::sys::thread::unix::Thread::new::thread_start::hc167e912ac26cc71
  57:     0x7f167de969cb - <unknown>
  58:     0x7f167df1aa0c - <unknown>
  59:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.93.0-nightly (f15a7f385 2025-11-04) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [mir_built] building MIR for `main`
#1 [check_unsafety] unsafety-checking `main`
#2 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 3 previous errors; 2 warnings emitted

Some errors have detailed explanations: E0412, E0580.
For more information about an error, try `rustc --explain E0412`.

```

</p>
</details>

<!--
query stack:
#0 [mir_built] building MIR for `main`
#1 [check_unsafety] unsafety-checking `main`
#2 [analysis] running analysis passes on crate `mvce`

-->

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/mod.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/mod.rs
@@ -250,29 +250,30 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
         }
 
         let mut reported = None;
-
         for from_expansion in [false, true] {
             for (error, suppressed) in iter::zip(&errors, &is_suppressed) {
-                if !suppressed
-                    && error.obligation.cause.span.from_expansion() == from_expansion
-                    && !error.references_error()
-                {
-                    let guar = self.report_fulfillment_error(error);
-                    self.infcx.set_tainted_by_errors(guar);
-                    reported = Some(guar);
-                    // We want to ignore desugarings here: spans are equivalent even
-                    // if one is the result of a desugaring and the other is not.
-                    let mut span = error.obligation.cause.span;
-                    let expn_data = span.ctxt().outer_expn_data();
-                    if let ExpnKind::Desugaring(_) = expn_data.kind {
-                        span = expn_data.call_site;
+                if !suppressed && error.obligation.cause.span.from_expansion() == from_expansion {
+                    if !error.references_error() {
+                        let guar = self.report_fulfillment_error(error);
+                        self.infcx.set_tainted_by_errors(guar);
+                        reported = Some(guar);
+                        // We want to ignore desugarings here: spans are equivalent even
+                        // if one is the result of a desugaring and the other is not.
+                        let mut span = error.obligation.cause.span;
+                        let expn_data = span.ctxt().outer_expn_data();
+                        if let ExpnKind::Desugaring(_) = expn_data.kind {
+                            span = expn_data.call_site;
+                        }
+                        self.reported_trait_errors
+                            .borrow_mut()
+                            .entry(span)
+                            .or_insert_with(|| (vec![], guar))
+                            .0
+                            .push(error.obligation.as_goal());
                     }
-                    self.reported_trait_errors
-                        .borrow_mut()
-                        .entry(span)
-                        .or_insert_with(|| (vec![], guar))
-                        .0
-                        .push(error.obligation.as_goal());
+                    if let Some(guar) = self.dcx().has_errors() {
+                        self.infcx.set_tainted_by_errors(guar);
+                    }
                 }
             }
         }

--- a/tests/ui/enum-discriminant/ice-from-type-error-issue-148515.rs
+++ b/tests/ui/enum-discriminant/ice-from-type-error-issue-148515.rs
@@ -1,0 +1,18 @@
+//@ edition: 2024
+
+enum Test {
+    Value = -5 >> 1_usize,
+}
+
+fn test1(x: impl Iterator<Item = Foo>) {
+    //~^ ERROR cannot find type `Foo` in this scope
+    assert_eq!(Test::Value as u8, -3);
+}
+
+fn test2(_: impl Iterator<Item = Foo>) {
+    //~^ ERROR cannot find type `Foo` in this scope
+    0u8 == -3;
+    //~^ ERROR cannot apply unary operator `-` to type `u8`
+}
+
+fn main() {}

--- a/tests/ui/enum-discriminant/ice-from-type-error-issue-148515.stderr
+++ b/tests/ui/enum-discriminant/ice-from-type-error-issue-148515.stderr
@@ -1,0 +1,24 @@
+error[E0412]: cannot find type `Foo` in this scope
+  --> $DIR/ice-from-type-error-issue-148515.rs:7:34
+   |
+LL | fn test1(x: impl Iterator<Item = Foo>) {
+   |                                  ^^^ not found in this scope
+
+error[E0412]: cannot find type `Foo` in this scope
+  --> $DIR/ice-from-type-error-issue-148515.rs:12:34
+   |
+LL | fn test2(_: impl Iterator<Item = Foo>) {
+   |                                  ^^^ not found in this scope
+
+error[E0600]: cannot apply unary operator `-` to type `u8`
+  --> $DIR/ice-from-type-error-issue-148515.rs:14:12
+   |
+LL |     0u8 == -3;
+   |            ^^ cannot apply unary operator `-`
+   |
+   = note: unsigned values cannot be negated
+
+error: aborting due to 3 previous errors
+
+Some errors have detailed explanations: E0412, E0600.
+For more information about an error, try `rustc --explain E0412`.

